### ISSUE TITLE
exception on update to 0.9.13

### ISSUE THREAD

**Issue by biancalana** [graphite-project/carbon#364]
I`ve just update to version 0.9.13  (getting https://github.com/graphite-project/carbon/tree/0.9.13-pre1) and now I`m getting the following error as soon my cache is full...

19/01/2015 18:46:42 :: MetricCache is full: self.size=1002173
19/01/2015 18:46:42 :: Exception in metricReceived event handler: args=('prod.vmware.myhost.net.vmnic7.sent_bps', (1421700280.0, 0.0)) kwargs={}
Traceback (most recent call last):
  File "/usr/lib64/python2.6/site-packages/twisted/internet/tcp.py", line 199, in doRead
    rval = self.protocol.dataReceived(data)
  File "/usr/lib64/python2.6/site-packages/twisted/protocols/basic.py", line 755, in dataReceived
    self.stringReceived(packet)
  File "/opt/graphite/lib/carbon/protocols.py", line 117, in stringReceived
    self.metricReceived(metric, datapoint)
  File "/opt/graphite/lib/carbon/protocols.py", line 64, in metricReceived
    events.metricReceived(metric, datapoint)
--- <exception caught here> ---
  File "/opt/graphite/lib/carbon/events.py", line 20, in __call__
    handler(*args, **kwargs)
  File "/opt/graphite/lib/carbon/cache.py", line 48, in store
    state.events.cacheFull()
exceptions.AttributeError: 'module' object has no attribute 'cacheFull'

I don`t think that is something that can be safely ignored....

**Comment by biancalana** [comment]
I'm not a python programmer, after do the following changes the error has disappeared

--- cache.py    2014-12-31 14:50:50.000000000 -0200
+++ /opt/graphite/lib/carbon/cache.py   2015-01-19 20:05:26.907780945 -0200
@@ -20,6 +20,7 @@
 except ImportError:
     from util import defaultdict

+from carbon import log, events, state, management

 class _MetricCache(defaultdict):
   def __init__(self, defaultfactory=deque, method="sorted"):
@@ -45,7 +46,7 @@
     self[metric].append(datapoint)
     if self.isFull():
       log.msg("MetricCache is full: self.size=%d" % self.size)
-      state.events.cacheFull()
+      events.cacheFull()

   def isFull(self):
     # Short circuit this test if there is no max cache size, then we don't need

**Comment by obfuscurity** [comment]
Actually, this should have been fixed in https://github.com/graphite-project/carbon/pull/351. Can you revert your change and upgrade to 2cc77752383288c27c4623a659cd0a76ef95b5a9 to see if that fixes it for you?

**Comment by biancalana** [comment]
yeah ! That fixed!

**Comment by obfuscurity** [comment]
Awesome, thanks for the update. `0.9.13` will definitely include that fix.

**Comment by cburroughs** [comment]
I'm seeing this today with carbon-realy after an upgade to 0.9.13

```
04/02/2015 08:56:05 :: Unhandled Error
Traceback (most recent call last):
  File "/opt/graphite/lib/carbon/client.py", line 267, in sendDatapoint
    self.client_factories[destination].sendDatapoint(metric, datapoint)
  File "/opt/graphite/lib/carbon/client.py", line 166, in sendDatapoint
    self.queueFull.callback(queueSize)
  File "/opt/graphite/lib/python2.6/site-packages/twisted/internet/defer.py", line 380, in callback
    self._startRunCallbacks(result)
  File "/opt/graphite/lib/python2.6/site-packages/twisted/internet/defer.py", line 488, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/opt/graphite/lib/python2.6/site-packages/twisted/internet/defer.py", line 575, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/opt/graphite/lib/carbon/client.py", line 113, in queueFullCallback
    state.events.cacheFull()
exceptions.AttributeError: 'module' object has no attribute 'cacheFull'

04/02/2015 08:56:06 :: Starting factory CarbonClientFactory(127.0.0.1:2014:None)
```

**Comment by deniszh** [comment]
@cburroughs - 0.9.13 is not released yet. Fix is not in 0.9.13-pre1, only in github repo AFAIK.

**Comment by obfuscurity** [comment]
Correct, the fix is in 0.9.x.

Jason

On Wed, Feb 04, 2015 at 10:03:59AM -0800, Denis Zhdanov wrote:
> @cburroughs - 0.9.13 is not released yet. Fix is not in 0.9.13-pre1, only in github repo AFAIK.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/graphite-project/carbon/issues/364#issuecomment-72905467

-- 
Jason Dixon
http://obfuscurity.com/
https://twitter.com/obfuscurity

**Comment by cburroughs** [comment]
Well as a user I find that confusing  since https://pypi.python.org/pypi/carbon/ has https://pypi.python.org/pypi/carbon/ has 0.9.13 and http://graphite.readthedocs.org/en/latest/releases/0_9_13.html says 'Graphite 0.9.13 is now available for usage'

**Comment by maroz** [comment]
Got the same issue, but in different place
Version: 0.9.13

```
Traceback (most recent call last):
  File "/opt/graphite/lib/carbon/client.py", line 267, in sendDatapoint
    self.client_factories[destination].sendDatapoint(metric, datapoint)
  File "/opt/graphite/lib/carbon/client.py", line 166, in sendDatapoint
    self.queueFull.callback(queueSize)
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 383, in callback
    self._startRunCallbacks(result)
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 491, in _startRunCallbacks
    self._runCallbacks()
--- <exception caught here> ---
  File "/usr/local/lib/python2.7/dist-packages/twisted/internet/defer.py", line 578, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/opt/graphite/lib/carbon/client.py", line 113, in queueFullCallback
    state.events.cacheFull()
exceptions.AttributeError: 'module' object has no attribute 'cacheFull'
```

**Comment by alexmuller** [comment]
We've upgraded Graphite to 0.9.13 and are seeing the same issue with our aggregation being broken. Is it possible to release a bugfix version that includes #351?

**Comment by obfuscurity** [comment]
@alexmuller Sorry for the confusion, but the "0.9.13" you're likely running is a release candidate. We intend to release a 0.9.13 (or 0.9.14, to alleviate any upgrade issues) soon.

**Comment by obfuscurity** [comment]
To clarify, that release **will** have the fix you're looking for, since it's already merged into the 0.9.x branch.

**Comment by alexmuller** [comment]
Thanks, 0.9.14 would be great - the package we installed is https://pypi.python.org/pypi/carbon/0.9.13.
